File: pascal/classes/class_encapsulation.pas
{$MODE OBJFPC}

program ClassEncapsulation;

type
  Account = class
  is_active: boolean; // public by default
  const
    connections_max_number = 10;
  public
    login: string;
    email: string;
    id: integer;
  protected
    connections: array[0..(connections_max_number - 1)] of integer;
  private
    connections_number: integer;
  public
    constructor create();

    procedure display();

    function hasConnections(): boolean;
    function addConnection(connected_account_id: integer): boolean;
  end;

  SocialMediaAccount = class(Account)
  public
    function isFriend(checking_account_id: integer): boolean;
  end;

constructor Account.create();
begin
  login := '';
  email := '';
  is_active := false;
  id := 0;
  connections_number := 0;
end;

procedure Account.display();
begin
  writeln('ID: ', id);
  writeln('Login: ', login);
  writeln('E-mail: ', email);
  writeln('Is active: ', is_active);
  writeln('Has connections: ', hasConnections());
end;

function Account.hasConnections(): boolean;
begin
  hasConnections := (connections_number > 0);
end;

function Account.addConnection(connected_account_id: integer): boolean;
begin
  if (connections_number = connections_max_number) then
      addConnection := false
  else
  begin
    connections[connections_number] := connected_account_id;
    inc(connections_number);
    addConnection := true;
  end;
end;

function SocialMediaAccount.isFriend(checking_account_id: integer): boolean;
var
  i: integer;
  connection_account_id: integer = 0;
begin
  for i := 0 to (connections_max_number - 1) do
  begin
    connection_account_id := connections[i];
    if (checking_account_id = connection_account_id) then
    begin
      isFriend := true;
      exit;
    end;
  end;

  isFriend := false;
end;

var
  timothy: SocialMediaAccount;

begin
  timothy := SocialMediaAccount.create();

  timothy.login := 'tim';
  timothy.email := 'timothy.muppetone@gmail.com';
  timothy.is_active := true;

  timothy.display();

  writeln();

  timothy.addConnection(100);
  writeln('Has connections: ', timothy.hasConnections());
  writeln('Is ID 100 a friend? ', timothy.isFriend(100));
  writeln('Is ID 200 a friend? ', timothy.isFriend(200));
end.

File: c++/classes/class_encapsulation.cpp
#include <iostream>

class Account {
  bool is_active = false; // private by default
public:
  std::string login = "";
  std::string email = "";
  unsigned int id = 0;
protected:
  const int connections_max_number = 10;
  /* For being used in the array definition
  const component must be also static. */
  int connections[10];
private:
  unsigned short int connections_number = 0;
public:
  void display();

  bool hasConnections() {
    return (connections_number > 0);
  };
  bool addConnection(unsigned int connected_account_id) {
    if (connections_number == connections_max_number)
      return false;

    connections[connections_number] = connected_account_id;
    connections_number++;

    return true;
  }
};

void Account::display() {
  std::cout << std::boolalpha
    << "ID: " << id << std::endl
    << "Login: " << login << std::endl
    << "E-mail: " << email << std::endl
    << "Is active: " << is_active << std::endl
    << "Has connections: " << hasConnections() << std::endl
    << std::noboolalpha;
}

class SocialMediaAccount : public Account {
public:
  bool isFriend(unsigned int checking_account_id) {
    for(int connection_account_id : connections) {
      if (checking_account_id == connection_account_id)
        return true;
    }

    return false;
  }
};

int main() {
  SocialMediaAccount timothy;

  timothy.login = "tim";
  timothy.email = "timothy.muppetone@gmail.com";
  // timothy.is_active = true; // cannot access private component

  timothy.display();

  std::cout << std::endl << std::boolalpha;

  timothy.addConnection(100);
  std::cout << "Has connections: " << timothy.hasConnections() << std::endl;
  std::cout << "Is ID 100 a friend? " << timothy.isFriend(100) << std::endl;
  std::cout << "Is ID 200 a friend? " << timothy.isFriend(200) << std::endl;

  std::cout << std::noboolalpha;
}
